Hlahd

riederd opened the pull request:
<!--
# nf-core/hlatyping pull request

Many thanks for contributing to nf-core/hlatyping!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/hlatyping/tree/master/.github/CONTRIBUTING.md)
-->

## Description

This PR adds support for HLA-HD (HLA typing from High-quality Dictionary), an algorithm known for providing accurate Class I and II HLA genotyping predictions with up to 6-digit (3-field) precision from WGS, WES, and RNA-Seq data (see also #194).

### Implementation Details

Due to licensing restrictions that require users to register and download the software directly from the provider, HLA-HD cannot be distributed via Conda environments or container images within this pipeline.

To address this, the integration supports a local installation workflow.

### Usage

Running HLA-HD is optional. Users must verify they have a valid local installation and enable the step using the following parameters:

* `--run_hlahd true`: Activates the HLA-HD process.

* `--hlahd_directory <path>`: Specifies the path to the local HLA-HD installation directory.

Example:
```
nextflow run nf-core/hlatyping \
    --input testing/samplesheet.csv \
    --outdir results \
    --run_hlahd true \
    --hlahd_directory /path/to/local/hlahd_v1.7.1
```

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/hlatyping/tree/master/.github/CONTRIBUTING.md)
- [x] Make sure your code lints (`nf-core pipelines lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

riederd committed:
Adds support for HLA-HD class II typing module as mentioned in #194

Integrates HLA-HD as an optional module for class II HLA typing, including configuration options, schema extension, environment setup, process definition, documentation, and unit tests. Enables users to run HLA-HD if a local installation is available, improving pipeline versatility for class II HLA genotyping.

riederd committed:
Updates output directory structure and version extraction

Organizes output files by sample ID for improved clarity and modifies the version parsing regex to correctly extract version numbers.
Removes unnecessary output directory creation step.

riederd committed:
Update changelog, citations, readme
fix hlahd publishDir

riederd committed:
Adds HLA-HD documentation to output guide

Introduces a new section detailing HLA-HD genotyping,
usage requirements, and output format to inform users
about optional HLA-HD support and integration steps.
Clarifies licensing, installation prerequisites, and
provides example output to enhance user understanding.

riederd committed:
cleanup modules.config

riederd committed:
Updates pipeline metadata, badges, and description

riederd committed:
fix linitng

riederd committed:
fix container options

riederd committed:
Fix container options handling for HLAHD module

riederd committed:
Warns and skips HLAHD typing if directory missing

Adds a check for the existence of the specified HLAHD installation directory.
If the directory is missing, logs a warning and skips HLAHD typing to prevent execution errors and clarify workflow behavior.

riederd committed:
Reorganizes paired check module file structure

Moves the paired check process into a subdirectory to improve
module organization and align with project structure conventions.

Fixes linting

riederd committed:
Corrects step numbering in usage instructions

Fixes a numbering error in the pipeline usage steps to maintain
accurate and sequential documentation for users.

riederd committed:
Updates HLA-HD module metadata and I/O definitions

Improves the module description and keywords for clarity, provides complete tool metadata including homepage, documentation, and license, and updates input/output specs to use FastQ files and result files. Enhances accuracy and compliance with current module standards.

riederd committed:
fix ro-create-metadata.json

christopher-mohr commented:
Cool, thank you for the contribution 👍 I will also have a look during this week.

jonasscheid commented on workflows/hlatyping.nf:
```diff
@@ -160,6 +161,20 @@ workflow HLATYPING {
+    //
+    // MODULE: Run HLAHD typing
+    //
+    if (params.run_hlahd) {
```
Design question (also to @christopher-mohr). With more tools incoming, should we switch to a new flag `--tools optitype,hlahd` analogue to epitopeprediction?

jonasscheid replied:
We should also document which version of HLA-HD is supported in the docs. I would vote for supporting the latest version only, since input params of the bash script might change in the future.

riederd replied:
* A `--tools` flag, with a default set,  would certainly be helpful
* supporting only the latest version is fine with me

jonasscheid commented on conf/modules.config:
```diff
@@ -36,15 +36,37 @@ process {
+                params.hlahd_directory ? "-f ${params.hlahd_directory}/freq_data" : '',
+            ].join(' ')
+        }
+        ext.args2 = {
```
Can ext.args2 also go into ext.args?

riederd replied:
I'm afraid no, because it seems that some of the order of the arguments for the `hlahd.sh` script is fixed.
* `ext.args` are options
* `in_reads` are positional arguments
* `ext.args2` are positional arguments (right after `in_reads`)

jonasscheid commented on conf/modules.config:
```diff
@@ -36,15 +36,37 @@ process {
 
-    withName: 'CAT_FASTQ' {
+    withName: HLAHD {
+        ext.path = params.hlahd_directory ? "${params.hlahd_directory}/bin" : ''
```
Could ext.path also be part of ext.args?

riederd replied:
I use `ext.path` to extend the system path `$PATH` in the `HLAHD` process. It is not an argument or option for the `hlahd.sh` bash script, however the script calls tools from the installation directory which need be found in `$PATH`.
Theoretically one could parse it out of `ext.args`, put in my opinion it just adds complexity.

We could, however, remove it from the `modules.config` and use `$params.hlahd_directory` directly in the `HLAHD` process to build the path.

jonasscheid replied:
I would prefer the latter solution: Give the module the path as an input param

jonasscheid replied:
One thing I am wondering is: Is it feasible to install HLA-HD within a pipeline run given the directory path? (Following the [installation doc](https://w3.genome.med.kyoto-u.ac.jp/HLA-HD/demo/hlahd-demo.v1.pdf) it might)

As context: We do this in epitopeprediction and create a folder with the ready-to-use executable as output of an "installation" module: https://github.com/nf-core/epitopeprediction/blob/master/modules/local/unpack_netmhc_software/main.nf

With this solution, the user is not forced to do the local installation. Not sure if that is feasible here though

riederd replied:
- I'll change the `ext.path` logic

- As for providing an installation module/process within the pipeline, it might be doable, however it implies compiling C code using `gcc` and running `bowtie2` for building. So we need either a container image with both the tools (is there one with gcc and bowtie2?) or 2 separate processes. 

I just quickly checked the installation module in the `epitopeprediction`. It seems that for each run (different users, different projects, deleted or different `work` dir) of the pipeline a new installation would be required. Is this correct? 

 I can look into it if you think it is worth it

riederd replied:
Ok, I implemented a similar installation process as you suggested. And refactored a bit.

* now only the path to the tarball of `hlahd` and `--run_hlahd` is needed as parameters
* there is now only one `ext.arg`
* it is also possible to let the user update to the latest hla dict
* I build a new container (docker, singularity) on sequera which have all dependencies installed

Thanks for the suggestions, looks better now :-)

jonasscheid replied:
Great! I'll have a look again

jonasscheid commented on modules/local/hlahd/main.nf:
```diff
@@ -0,0 +1,71 @@
+        ? 'https://community-cr-prod.seqera.io/docker/registry/v2/blobs/sha256/b4/b41b403e81883126c3227fc45840015538e8e2212f13abc9ae84e4b98891d51c/data'
+        : 'community.wave.seqera.io/library/bowtie2_htslib_samtools_pigz:edeb13799090a2a6'}"
+
+    containerOptions "${workflow.containerEngine == 'singularity' && task.ext.containerOptions ? "--bind ${task.ext.containerOptions}" : "--volumes ${task.ext.containerOptions}"}"
```
Could containerOptions also live in modules.config?

riederd replied:
Did not try to. I wanted to hide it from the configuration to improve its readability.

jonasscheid replied:
I think its fine, only personal touch. Since it will remain a local module most likely, that should be fine as well

jonasscheid reviewed (changes requested):
Thanks for the contribution of HLA-HD! This will definitevly be a benefit for this pipeline :)

I raised a couple of design questions (also towards main developer @christopher-mohr), but nothing major. One thing that still is missing is the nf-test test of this new module.

riederd committed:
Change HLAHD path resolution logic

Get rid of ext.path and compile the hlahd path from params.hlahd_directory in the module process

riederd committed:
Refactors HLA-HD handling to support automated installation

Replaces directory-based HLA-HD usage with a tarball-based workflow,
automating software installation and optional dictionary updates.
Introduces process separation for installation and genotype steps,
updates configuration parameters and schema, and streamlines
argument handling for reproducible, user-friendly execution.

riederd committed:
Aligns YAML indentation for conda environment files

Updates indentation in environment configuration files to conform with YAML standards, improving readability and ensuring compatibility with YAML parsers.

riederd committed:
Adds YAML document start marker for schema compliance

Introduces the '---' YAML document start in environment files
to improve compatibility with schema validation tools and
clarify document structure.

jonasscheid commented on modules/local/hlahd/genotype/main.nf:
```diff
@@ -20,9 +18,12 @@ process HLAHD {
 
     script:
-    def hlahd_p = file(params.hlahd_directory).isDirectory() ? "${params.hlahd_directory}/bin" : ''
+    def hlahd_p = hlahd_directory ? "${hlahd_directory}/bin" : ''
```
I guess the tool is not functional if you dont provide the hlahd_directory right? It might make sense to get rid of the logic here and throw an error if hlahd is specified but the directory is not specified. Could be part of the workflow-script

riederd replied:
True. I'll change it.

jonasscheid commented on modules/local/hlahd/install/main.nf:
```diff
@@ -0,0 +1,66 @@
+        sh update_dictionary.sh
+    fi
+
+    cd ../../
```
Needed or dev-artifact?

riederd replied:
needed, so that the `versions.yml` is generated correctly

jonasscheid commented on nextflow_schema.json:
```diff
@@ -87,12 +87,19 @@
+                    "help_text": "Path to the installation tarball of HLA-HD. You need to provide this tarball to run HLA-HD. Obtain a licensed copy from https://w3.genome.med.kyoto-u.ac.jp/HLA-HD/"
+                },
+                "hlahd_update_dictionary": {
+                    "type": "boolean",
```
Still in favor of only supporting one version. That means we do not need the update flag right?

riederd replied:
The idea here is to allow users to update the IPD-IMGT/HLA database to include "new" HLA types. This does not update the software itself. 

Do you foresee any potential issues with that? I think it's similar to providing another genome reference in other pipelines.

jonasscheid replied:
Ahh yes of course, good point! I'll cross mention https://github.com/nf-core/hlatyping/issues/197 here. Maybe we can do something similar as you did with optitype in the future

jonasscheid replied:
Maybe we can still change the flag to `hlahd_update_hla_reference` or `hlahd_update_reference`

riederd replied:
Hmm. Given that the "HLA type reference" is termed as "dictionary" throughout the HLA-HD docs, I'm not sure if we should avoid "dict" or "dictionary" in the flag name.

What about: `hlahd_update_reference_dict`?

jonasscheid replied:
👍🏼

riederd committed:
Adds validation for missing HLAHD directory

Introduces an explicit error when the specified directory does not exist,
improving early failure detection and preventing downstream errors due to
invalid input paths.

riederd committed:
Renames HLA-HD dictionary update param for clarity

Updates parameter and related docs to use a more descriptive name for the HLA-HD reference dictionary update option, clarifying its association with the IGMT/HLA database and improving overall configuration readability.

jonasscheid commented on docs/output.md:
```diff
@@ -57,6 +58,25 @@ Additionally, a coverage plot of the predicted HLA genotype is produced for qual
+[HLA-HD](https://w3.genome.med.kyoto-u.ac.jp/HLA-HD/) (HLA typing from High-quality Dictionary) is an HLA genotyping algorithm that utilizes an extensive dictionary of HLA alleles and a unique mapping strategy. It calculates scores based on weighted read counts to select the most suitable allele pairs. Unlike many other tools, HLA-HD can provide accurate HLA genotyping predictions with up to 6-di ...
+For further reading and documentation see the [HLA-HD website](https://w3.genome.med.kyoto-u.ac.jp/HLA-HD/).
+
+> **Note:** To run HLA-HD within the nf-core hlatyping pipeline, a local installation of HLA-HD is required. HLA-HD is not distributed with the pipeline's containers due to licensing restrictions. The software is freely available for academic and non-commercial research purposes, but users must register and download it directly from the provider. Commercial use requires a specific license.
```
```suggestion
> **Note:** HLA-HD is not distributed with the pipeline's containers due to licensing restrictions. The software is freely available for academic and non-commercial research purposes, but users must register and download it directly from the provider. Commercial use requires a specific license.
```

jonasscheid commented:
Hi @riederd, Thanks again for the PR! From my side only two things are missing: 
- Switch to e.g. `--tools optitype,hlahd` makes sense (optitype only should be default).
- Add test_hlahd.config in `conf` and add an nf-test script with snapshot in `tests` folder

riederd committed:
Update docs/output.md

Remove requirement for local/manual installation of HLA-HD

Co-authored-by: Jonas Scheid <43858870+jonasscheid@users.noreply.github.com>

riederd commented:
> Hi @riederd, Thanks again for the PR! From my side only two things are missing:
> 
> * Switch to e.g. `--tools optitype,hlahd` makes sense (optitype only should be default).
> * Add test_hlahd.config in `conf` and add an nf-test script with snapshot in `tests` folder

I'll start working on this

riederd committed:
Adds multi-tool HLA typing selection via parameter

Introduces a unified parameter for specifying which HLA typing tools to run, allowing flexible selection between OptiType and HLA-HD.
Removes legacy single-tool configuration and updates workflow logic to validate tool choices and run selected modules conditionally.
Improves extensibility and user control over HLA typing approaches.

riederd committed:
fix schema

riederd committed:
Fix using tools option
Remove run_hlahd option left over
Fix hlahd_directory check

riederd committed:
Improves HLAHD path validation logic

Handles both null and non-existent HLAHD path values to prevent workflow errors when the path is not provided or invalid.
Enhances user guidance by warning for missing parameter input.

riederd committed:
Adds test profiles and unit tests for HLA-HD and Optitype

Introduces dedicated Nextflow configuration profiles to support HLA-HD and combined Optitype+HLA-HD test runs, along with corresponding unit tests for improved pipeline validation.

Expands contributor list and updates documentation to reflect new test capabilities.

riederd committed:
Adds missing newline at end of config files

riederd committed:
add test snapshots
update .gitignore

riederd commented:
> > Hi @riederd, Thanks again for the PR! From my side only two things are missing:
> > 
> > * Switch to e.g. `--tools optitype,hlahd` makes sense (optitype only should be default).
> > * Add test_hlahd.config in `conf` and add an nf-test script with snapshot in `tests` folder
> 
> I'll start working on this

* implemented `--tools` option with default `optitype? enabled
* added tests for `hlahd`

the tests work locally, however they fail on the github-ci. I guess this is due to the missing hlahd tar.gz installation file, so I think ci-tests on github will never run successful. What do you suggest to solve this @jonasscheid ?

jonasscheid commented on CHANGELOG.md:
```diff
@@ -12,6 +12,10 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
+### `Added`
+
+- Option to run HLA-HD for Class I and Class II HLA typing
```
```suggestion
- [#201](https://github.com/nf-core/hlatyping/pull/201) Option to run HLA-HD for Class I and Class II HLA typing (@riederd)
```

jonasscheid committed:
Apply suggestion from @jonasscheid

jonasscheid commented:
> the tests work locally, however they fail on the github-ci. I guess this is due to the missing hlahd tar.gz installation file, so I think ci-tests on github will never run successful. What do you suggest to solve this @jonasscheid ?

Well done :) Thats right! So now the fun part starts:
In order to test a licensed tool, we need to encrypt it, put it on nf-core/testdata and when the ci is run we need to decrypt it again. We need to give the encryption key as a github secret BUT github secrets are only propagated in PRs WITHIN the repo, meaning the test will always fail if the PR comes from a fork. I'm going to do the encryption part and put it into the testdata repo meanwhile you can go ahead and check how the ci is done in epitopeprediction with the licensed netmhcpan tool. I'll let you know when you can start opening the fix-ci branch from within this repo. This PR can be merged now. Thanks!! 🚀

jonasscheid commented:
@riederd The encrypted file is now uploaded under https://github.com/nf-core/test-datasets/blob/hlatyping/software/hlahd.1.7.1.tar.gz.gpg. Could you fullfill the merge requirements of git to merge this one?

riederd commented:
> @riederd The encrypted file is now uploaded under https://github.com/nf-core/test-datasets/blob/hlatyping/software/hlahd.1.7.1.tar.gz.gpg. Could you fullfill the merge requirements of git to merge this one?

@jonasscheid Thanks! However, I'm afraid I need some guidance to get this done. Sorry about this but I'm a bit lost.

riederd committed:
Update configuration files and revert changes in pull request template

riederd committed:
lint and fix

riederd committed:
revert git ignore

riederd committed:
update nf core version to fix linting

riederd committed:
Revert to nf-core 3.3.2

riederd committed:
fix linting

riederd committed:
add lint section to .nf-core.yml for unchanged files

prevents linting failures on github, i.e.:
  - docs/images/nf-core-hlatyping_logo_dark.png does not exist in the dev branch
  - assets/nf-core-hlatyping_logo_light.png was not changed by @riederd
  - .github/PULL_REQUEST_TEMPLATE.md was not changed by @riederd

riederd commented:
@jonasscheid  I "fulfilled" the merge requirements of git  by making some exceptions from the "files_unchanged" linting:

      - docs/images/nf-core-hlatyping_logo_dark.png does not exist in the dev branch
      - assets/nf-core-hlatyping_logo_light.png from dev was not changed by @riederd
      - .github/PULL_REQUEST_TEMPLATE.md from dev was not changed by @riederd
      
I'm not sure if this was the correct way to get over git check failures, but I could not figure out an alternative.
What are now the next steps? Just click "Merge pull request" ?

jonasscheid commented:
Great thanks! Two tests are still failing, could you fix those? You might need to run `nf-test test tests --profile=+docker -update-snapshot` for this

EDIT: Just reminded myself again that we need to fix the hlahd tests with the follow-up PR. Ok I'll merge it

jonasscheid commented:
>What are now the next steps? J

Next step would be a feature branch within the nf-core/hlatyping repo to implement the CI tests for hlahd. Since time is precious currently, I'm not sure if I will be able to do it this week. Let me know if you need to have it this week or if you are fine as is

riederd commented:
Thanks! I can wait, no hurries

jonasscheid commented:
@riederd Looks like tests are running on CI as well! 
Could you give me a review for #203. Then we can do a release
